documentation/documentation_publisher.ts

```diff
@@ -0,0 +1,110 @@
+import AWS from 'aws-sdk';
+import S3 from 'aws-sdk/clients/s3';
+import {promisify} from 'util';
+import {print, printAndExit} from '../testing/helpers';
+import {version} from '../package.json';
+import yargs, {Arguments} from 'yargs';
+import {exec as childExec} from 'child_process';
+const exec = promisify(childExec);
+
+const AwsRegion = 'us-west-2';
+const BaseGeneratedDocsPath = 'site';
+const DocumentationBucket = 'developer-documentation';
+const PacksSdkBucketRootPath = 'packs';
+const EnvironmentKeyRoot = `${PacksSdkBucketRootPath}/current`;
+const Environments = ['adhoc', 'head', 'staging', 'prod'];
+
+function handleError(e: Error) {
+  printAndExit(e.message + ` ${e.stack || ''}`);
+}
+
+function getS3Service(env: string): S3 {
+  const config: S3.ClientConfiguration = {
+    signatureVersion: 'v4',
+    region: AwsRegion,
+    credentials: new AWS.SharedIniFileCredentials({profile: env}),
+    computeChecksums: true,
+  };
+  return new S3(config);
+}
+
+function getS3Bucket(env: string): string {
+  return `coda-us-west-2-${env}-${DocumentationBucket}`;
+}
+
+function getS3ConfigKey(): string {
+  return `${PacksSdkBucketRootPath}/${version}`;
+}
+
+function getS3EnvironmentKey(environment: string): string {
+  return `${EnvironmentKeyRoot}/${environment}`;
+}
+
+async function pushDocsToEnv(env: string) {
+  const s3 = getS3Service(env);
+  const bucket = getS3Bucket(env);
+  const key = getS3ConfigKey();
+
+  async function pushDocsDirectory(): Promise<any> {
+    await exec(`aws s3 sync --profile ${env} --region ${AwsRegion} ${BaseGeneratedDocsPath} s3://${bucket}/${key}`);
+  }
+
+  try {
+    // If folder already exists, warn since we are uploading all the documentation wholesale instead of "syncing."
+    const obj = await s3.listObjectsV2({Bucket: bucket, MaxKeys: 1, Prefix: key}).promise();
+    if (!obj.Contents?.length) {
+      console.warn(`Folder already exists for ${version}.`);
+    }
+    print(`Pushing the current packs-sdk documentation ${key} to ${env}...`);
+    await pushDocsDirectory();
+    print(`The current packs-sdk documentation was pushed to ${env}:${key} successfully.`);
+  } catch (err: any) {
+    handleError(err);
+  }
+}
+
+async function pushDocumentation() {
+  await Promise.all(Environments.map(env => pushDocsToEnv(env)));
```

Comment:
Probably want these to get pushed in a deterministic order, and let one succeed before trying the others? In which case, could use spawnSync and just use a regular `for` loop.

`prod` needs to get split out into a separate step I think though, in order for CircleCI to halt and wait for approval first?

Reply:
sure that sounds reasonable for pushing them in order.

the reason they all get pushed at the same time (including prod) is because the runtime config flow is 1) push everything 2) progressively active the new version for each environment with activating the new prod waiting on approval. Will see how feasible we can use the "active" piece with Cloudfront serving and then revisit the flow here if not.

Reply:
although I guess if we're making it possible to serve all versions at once we actually want the push step to be hidden too since it might accidentally expose the latest version that hasn't been approved if someone is able to enumerate the URL so maybe just cleaner to only push once we approve too

Reply:
ok this is resolved with the new logic